### ISSUE TITLE
3.6 Multiple assertion backwards compatibility

### ISSUE THREAD

**Issue by AndyBan** [nunit/nunit#2007]
We use nunit along with selenium to check that elements are present on a web page etc.  Our framework uses exception capturing loops for certain cases such as interacting with browsers which isn't always the most reliable thing.  This allows a bit of time for elements to appear which they are created by Javascript however if things don't appear after a short time then the nunit assertion exception is not captured and bubbles up to the runner and is reported on.

We have found since upgrading to 3.6 that the failed assertions are getting recorded due to the new multiple assertion feature even though the same assertion later passes. I think exceptions should not be recorded unless you are inside a multiple assertion block, would you agree?

A potential fix is to move this earlier in the ReportFailure method:
if (TestExecutionContext.CurrentContext.MultipleAssertLevel == 0)
                throw new AssertionException(writer.ToString());

in src/NUnitFramework/framework/Assert.cs

**Comment by CharliePoole** [comment]
The future direction for NUnit is that all assertion failures will be recorded. For that reason, we will need to solve your problem in some other way than you suggest. Can you provide a bit more info (code or pseudo-code?) about how you are ignoring some failures and re-running the tests?

**Comment by stevenebutler** [comment]
We are using the same technique in our selenium based framework with the same issues on nunit 3.6.1.  We have reverted to using nunit 3.4.1 until we find a good solution for use with nunit 3.6.1.

I will try to outline what our framework does to synchronise on pages here.

We define a web page class that encapsulate actions on the page, as well as means to identify items on screen, including information sections, web page controls etc.  This page also has a standard way of waiting for a web page to load.  It uses a `SeleniumExtensions wait.Until()` construct to periodically poll an `AssertPageShowing()` method in the derived page class. 

The `AssertPageShowing()` method simply does things like:

```
AssertPageTitle("My page title");  // Convenience method that calls Assert.That() on the title of the browser
Assert.That(ControlIdentifier.Displayed, "Important control that signals the page has finished loading is present and able to be interacted with");
Assert.That(OtherControl.Displayed, "etc");
```
As the page is loading dynamically, the wait.Until() will continually trigger nunit assertions that we ignore during the timeout period.  Once the wait.Until() completes, we call the assertion function a final time to ensure that it did actually load the page successfully, before returning the page to the calling method to allow interactions to proceed.

We want the assertions to be ignored while the `wait.Until()` hasn't expired, but the assertions should be logged once the timeout has expired (and thus register a test failure only at that time).  The new multiple assertions feature in 3.6.1 is failing our tests even when we would have expected them to pass.

Note that we originally did not use assertions in the `AssertPageShowing()` method, and instead had an `IsPageShowing()` method that returned true or false.  The problem with this approach was we had no idea what caused the test to fail if there are multiple items that needed to be present, or in particular states, for us to consider the page load complete.  By using the assertions, we had instant feedback in the test logs as to why the page load failed.  

If there is another approach that would work better with nunit 3.6.1 I would consider changing our design, but otherwise, as per the original issue post, I'd really appreciate a way to stop assertions registering while we're waiting to see if the assertions will stop failing.

**Comment by JustinRChou** [comment]
Would it be possible to use a `DelayedConstraint`?

**Comment by CharliePoole** [comment]
The only issue with a `DelayedConstraint` is that it applies to a single assertion. However, if he can deal with repeating __each__ assertion until it passes, rather than __all__ assertions, I think that would work.

@stevenebutler Could you do something like this?

```C#
AssertThat(Page.Title, Is.EqualTo("Page Title").After(...).PollEvery(...));
Assert.That(ControlIdentifier.Displayed).After(...).PollEvery(...));
Assert.That(OtherControl.Displayed).After(...).PollEvery(...));
```

Alternatively, wrap them all into a single method call and use `After` on an assertion that calls that method.

**Comment by stevenebutler** [comment]
That might work, but up until now I didn't even realise the `DelayedAssertion` existed, so it'll take me some time to experiment.

**Comment by JustinRChou** [comment]
I was experimenting using the `DelayedConstraint` by changing a string in a thread. Had to use a StringBuilder.
```
        [Test]
        public void DelayTest()
        {
            Thread th = new Thread(()=> {
                Thread.Sleep(2000);
                sb.Append("Added");
            });
            th.Start();
            Assert.That(sb, Is.EqualTo(new StringBuilder("StartAdded")).After(5000));
            th.Join();
        }
```
The Assert fails which is probably failing on the reference. I thought `EqualTo` didn't compare the reference?

**Comment by CharliePoole** [comment]
NUnit has no special equality processing for `StringBuilder` so the `StringBuilder.Equals` override is used. Equality for a `StringBuilder` is defined as having the same content, capacity and max capacity. IOW, it's not very useful to compare `StringBuilder`s.

This would work, however

```C#
Assert.That(() => sb.ToString(),
```

**Comment by JustinRChou** [comment]
That indeed worked.

**Comment by CharliePoole** [comment]
@AndyBan Did you want this open or closed? 😄

**Comment by AndyBan** [comment]
I was a bit early on reopening it, a bit of further investigation and I found things are ok :)

Whilst the above works for us it is a little bit verbose would be my only comment.  We also configure our waits globally as we found it wasn't very maintainable to configure them on a per instance basis.  I've yet to confirm with my team on the best solution but its looking like we'll handle the waiting for the answer before calling through to your assert as selenium throws exceptions which wouldn't be handled by your interval assert which I think is correct behaviour.

**Comment by stevenebutler** [comment]
>@stevenebutler Could you do something like this?

>``AssertThat(Page.Title, Is.EqualTo("Page Title").After(...).PollEvery(...));
> Assert.That(ControlIdentifier.Displayed).After(...).PollEvery(...));
> Assert.That(OtherControl.Displayed).After(...).PollEvery(...));``

I don't think this approach would work for me, since we want the .After() to apply to the group of tests as a whole.

> Alternatively, wrap them all into a single method call and use After on an assertion that calls that method.

This seems like the way to go, but I'm not sure how to express it in nunit.

``Assert.That(() => AssertPageShowing().After(...).PollEvery(...)));``

doesn't work, since AssertPageShowing isn't a constraint.  Would this approach require providing some form of constraint interface to the method?

I had misread the problem I was having though.  It turns out I don't get the polling failures logged in total unless the overall test fails anyway.  If the polling eventually succeeds, my log is not polluted with the retry attempts, only if the test failed anyway.

**Comment by JustinRChou** [comment]
I wonder if you could do your checks in another method that returns a boolean and do something like
`Assert.That(TestOnPage(), Is.EqualTo(true).After(timeout));`

**Comment by stevenebutler** [comment]
In my original comment I mentioned I'd previously used an IsShowing() method, but due to the problem of identifying the cause of failures, I ultimately changed to the approach I use now with asserting each issue.

To that end, I tried your suggestion, with assertions in the method, but unfortunately, it ends fails on the first assertion.  It seems like it wasn't designed to be used in this way, which is understandable.

```
        int x = 0;
        private bool IsShowing()
        {
            x++;
            Console.WriteLine("Polled at "+x);
            Assert.That(x, Is.GreaterThan(50), "Message 50");
            Assert.That(x, Is.GreaterThan(51), "Message 51");
            Assert.That(x, Is.GreaterThan(52), "Message 52");
            return true;
        }
        [Test]
        public void DelayedAssertTest()
        {
            Assert.That(() => IsShowing(), Is.True.After(3000).PollEvery(10));
            
        }
```

A second attempt in which I tried to treat assertion failures as true/false returns looks like this, but it ultimately fails because all the assertion failures in the After() are collected and treated as test failure at the end of the test.  Also, if it didn't collect the failures and the conditions weren't met, I'd still be missing my failure reason at the end.

```     
        private bool IsShowing2()
        {
            x++;
            Console.WriteLine("Polled at "+x);
            try
            {
                Assert.That(x, Is.GreaterThan(50), "Message 50");
                Assert.That(x, Is.GreaterThan(51), "Message 51");
                Assert.That(x, Is.GreaterThan(52), "Message 52");
            } catch (Exception)
            {
                return false;
            }
            return true;
        }
        [Test]
        public void DelayedAssert2Test()
        {
            x = 0;
            Assert.That(() => IsShowing2(), Is.True.After(3000).PollEvery(10));

        }
```

**Comment by JustinRChou** [comment]
I was thinking something like this 
```
        int x = 0, y= 0 ,z = 0  ;
        private bool IsShowing()
        {
            x++;y++;z++;
            bool result = false;
            if (x >= 50 && y>=50 && z >= 50) {
                result = true;
            }
            return result;
        }
        [Test]
        public void DelayedAssertTest()
        {
            Assert.That(() => IsShowing(), Is.True.After(3000).PollEvery(10));
            Console.WriteLine(x);
        }
```
But you get a very uninformative assert failure.

**Comment by stevenebutler** [comment]
Yes, the uninformative assert failure is what I'm trying to avoid.  Thanks for the suggestions though.

It looks like my worst symptom with my original approach (using selenium) is that on failure I get a potentially large list of failed exceptions in the nunit failure message (one from each failed poll).

**Comment by JustinRChou** [comment]
```
        int x = 0, y= 0 ,z = 0  ;
        string assertFailure = "";
        private bool IsShowing()
        {
            x++;y++;z++;
            bool result = false;
            assertFailure = "";
            if (x >= 50 && y>=50 && z >= 50) {
                result = true;
            }
            if (y <50) {
                assertFailure += "y less than 50";
            }
            if (z < 50)
            {
                assertFailure += "z less than 50";
            }
            return result;
        }
        [Test]
        public void DelayedAssertTest()
        {
            Assert.That(() => IsShowing(), Is.True.After(3000).PollEvery(1000), ()=> { return assertFailure; });
            Console.WriteLine(x);
        }
```
At least now this gives a better assert failure message.

**Comment by stevenebutler** [comment]
Yes, that would work, but it is significantly more code than the simple series of asserts we have now.  We have to write one of these methods for each page that we have to synchronise on in the application, so ease of creation is an important consideration.

I will have to think about the best way to handle this so that there is still a simple interface, but for now I will live with the extra logged assertions.

**Comment by stevenebutler** [comment]
This is a workaround I came up with that meets my needs, but does have the disadvantage of using code in the Internal namespace `TestExecutionContext.IsolatedContext` (probably exposed for testing nunit) to isolate the assertions that are being fired in the delay loop from the test result:

```
       int x = 0;
        private void IsShowing()
        {
            x++;
            Console.WriteLine("Polled at " + x);
            Assert.That(x, Is.GreaterThan(50), "Message 50");
            Assert.That(x, Is.GreaterThan(51), "Message 51");
            Assert.That(x, Is.GreaterThan(52), "Message 52");
        }
 
        private bool ExceptionToBool(TestDelegate del)
        {
            try
            {
                del();
                return true;
            }
            catch (Exception)
            {
                return false;
            }

        }

        [Test]
        public void DelayedAssertTest()
        {
            x = 0;
            using (new TestExecutionContext.IsolatedContext())
            {
                    Warn.If(ExceptionToBool(IsShowing), Is.True.After(3000).PollEvery(1000));
            }
            
            Assert.Multiple(IsShowing);

        }
```

I am leaning towards not using `Assert.Multiple` in the final step, because it ends up obscuring the failure assertion exceptions when running under the debugger, but the option is there to have it run all tests in the final block or only up to the first failure in the post time-out final check.

**Comment by CharliePoole** [comment]
Good on you for finding that. I've been looking at this thread and asking myself "Didn't I just have to do something to fix a similar problem?" Well, that was it. I use the IsolatedContext in the implementation of Throws because we might be asserting that an AssertionException is thrown!

It's in the Internal namespace for use both internally and by extensions. I considered providing something public to wrap it (Assert.Isolated?) but ended up not doing that.

**Comment by stevenebutler** [comment]
Ok, that's great that it's not meant to be completely off-limits.

It was a colleague of mine who found the use in the unit tests.  I had originally resorted to creating a temporary execution context inline while proving the concept, and then simplified the whole thing once he pointed out what the tests were doing.

**Comment by JustinRChou** [comment]
I can't seem to find docs on `TestExecutionContext`

**Comment by CharliePoole** [comment]
We don't document internal stuff generally, although that one may need to be in our extensibility docs when we get them done.

**Comment by stevenebutler** [comment]
I've noticed an issue when I use `Warn.If(...,Is.True.After(...).PollEvery(...)` it only polls once, at the end of the timeout.
I believe using Warn.If was a logical error and I should have used Warn.Unless(), which does poll at the specified rate.
```
            using (new TestExecutionContext.IsolatedContext())
            {
                    Warn.Unless(ExceptionToBool(IsShowing), Is.True.After(3000).PollEvery(1000));
            }
```  

  I'm assuming the logic on `Warn.If` to not poll until the end is by design based on a grammatical interpretation of the constraint?

**Comment by CharliePoole** [comment]
That's correct. The `Until` construct causes the test to be repeated until it succeeds or times out. However, under the control of `Warn.If`, the constraint provided is actually replaced with it's negation, i.e. `Is.Not.True...` and that replacement constraint "succeeds" immediately.

We should probably do something about this: either making it work as you expected or disallowing the use of a `DelayedConstraint` with `Warn.If`. What do folks think?

**Comment by rprouse** [comment]
I think a possible use of `Warn.If` would be to check for a connection to an external system or something similar in integration testing, so supporting the `DelayConstraint` makes sense.

**Comment by stevenebutler** [comment]
I had some trouble with my unit test case that I had used to analyze the problem (although my actual implementation is working because I used a function name instead of a wrapper function) and have finally realized that I needed to use a lambda expression.  Otherwise, `ExceptionToBool(IsShowing)` is calculated once only and the same value is repeatedly tested.   My comment above should have had

```
...
            using (new TestExecutionContext.IsolatedContext())
            {
                Warn.Unless(() => { return ExceptionToBool(IsShowing); }, Is.True.After(3000).PollEvery(100));
            }
...
```

Sorry if my test case confused anyone because it didn't actually work.  I need to make sure I rerun after I tidy things up next time.

**Comment by stevenebutler** [comment]
I am seeing something strange in the polling frequency.  I set it the poll frequency to 250 ms, but by adding some tracing to the IsShowing method, I can see that it is frequently double polling the delegate value.  Would you consider this a bug:
```
Test Name:	DelayedAssertTest
Test Outcome:	Failed
Result Message:	
Message 50
  Expected: greater than 50
  But was:  25
Result StandardOutput:	
Polled at 1 2017-02-01 08:20:05.689
Polled at 2 2017-02-01 08:20:05.702
Polled at 3 2017-02-01 08:20:05.953
Polled at 4 2017-02-01 08:20:05.953
Polled at 5 2017-02-01 08:20:06.204
Polled at 6 2017-02-01 08:20:06.204
Polled at 7 2017-02-01 08:20:06.456
Polled at 8 2017-02-01 08:20:06.456
Polled at 9 2017-02-01 08:20:06.708
Polled at 10 2017-02-01 08:20:06.708
Polled at 11 2017-02-01 08:20:06.958
Polled at 12 2017-02-01 08:20:06.958
Polled at 13 2017-02-01 08:20:07.210
Polled at 14 2017-02-01 08:20:07.210
Polled at 15 2017-02-01 08:20:07.461
Polled at 16 2017-02-01 08:20:07.461
Polled at 17 2017-02-01 08:20:07.712
Polled at 18 2017-02-01 08:20:07.712
Polled at 19 2017-02-01 08:20:07.963
Polled at 20 2017-02-01 08:20:07.966
Polled at 21 2017-02-01 08:20:08.215
Polled at 22 2017-02-01 08:20:08.217
Polled at 23 2017-02-01 08:20:08.438
Polled at 24 2017-02-01 08:20:08.438

```

**Comment by CharliePoole** [comment]
See issue #2043 and the proposed solution. I'm keeping several issues open for now, even though they may be duplicates, because each gives a slightly different use case.

**Comment by CharliePoole** [comment]
This was resolved in the 3.6.1 release
